File: 백준/Gold/2589. 보물섬/보물섬.py
import sys
input = sys.stdin.readline

from collections import deque

dxs = [1, 0, -1, 0]
dys = [0, 1, 0, -1]

def is_reachable(x, y):
    return 0 <= x < n and 0 <= y < m and (matrix[x][y] != "W")

def bfs(x, y):
    visited = [[-1] * m for _ in range(n)]

    visited[x][y] = 0
    queue = deque([[x, y]])

    cnt = 0

    while queue:
        x_, y_ = queue.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x_ + dx, y_ + dy

            if is_reachable(nx, ny) and visited[nx][ny] == -1:
                queue.append([nx, ny])
                visited[nx][ny] = visited[x_][y_] + 1
                cnt = max(cnt, visited[nx][ny])

    return cnt


n, m = map(int, input().split())

matrix = []

for _ in range(n):
    matrix.append(list(input()))

max_dist = -1

for x in range(n):
    for y in range(m):
        if matrix[x][y] != "W":
            cnt = bfs(x, y)
            max_dist = max(max_dist, cnt)

print(max_dist)

File: 백준/Gold/2589. 보물섬/보물섬.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, arr[51][51], visited[51][51], ans;
string temp;

const int dxs[] = {1, 0, -1, 0};
const int dys[] = {0, 1, 0, -1};

bool is_reachable(int x, int y) {
	return (0 <= x && x < n) && (0 <= y && y < m);
}

void bfs(int x, int y) {
	memset(visited, 0, sizeof(visited));
	
	queue<pair<int, int>> q;
	q.push({x, y});
	visited[x][y] = 1;

	while (q.size()) {
		tie(x, y) = q.front(); q.pop();
		ans = max(ans, visited[x][y]);

		for (int i = 0; i < 4; ++i) {
			int nx = x + dxs[i];
			int ny = y + dys[i];

			if (!is_reachable(nx, ny))	continue;
			if (visited[nx][ny])		continue;
			if (arr[nx][ny] == 0)		continue;

			visited[nx][ny] = visited[x][y] + 1;
			q.push({nx, ny});
		}
	}
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
	cin >> n >> m;

	for (int i = 0; i < n; ++i) {
		cin >> temp;
		for (int j = 0; j < m; ++j) {
			if (temp[j] == 'W')	arr[i][j] = 0;
			else				arr[i][j] = 1;
		}
	}

	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < m; ++j) {
			if (arr[i][j] == 1)	bfs(i, j);
		}
	}

	cout << ans - 1;

	return 0;
}